### ISSUE TITLE
Feature Request: Interactive workflows

### ISSUE THREAD

**Issue by syvsto** [onivim/oni2#681]
As discussed in the onivim Discord channel yesterday, integrating in-line code execution and other tools for interactive workflows would be helpful for us folks who often do exploratory programming. Whether it is due to the restrictions of rendering in a terminal or to the adherence to the UNIX philosophy, vim does not support interactive workflows well, especially when it comes to integrating prose, images, and code into the programming workflow. Oni2, with its proper GUI, does not suffer from the same limitations.

The feature requested will require quite a few man-hours of development and is non-essential for many day-to-day editing tasks. The goal of this issue is therefore not to push for immediate development, but rather to collect and discuss interesting approaches to implementing the feature, so any future development efforts are well-founded.

### Avenues to explore

There are multiple existing approaches to interactive programming environments. The following approaches are the ones I have experience with, and though there is considerable overlap between the two, I see them as different enough that they can be considered different idioms.  All descriptions are based on personal opinion (and too little actual research), so take everything with a grain of salt.

#### [Light Table](http://lighttable.com/)/[SLIME](https://common-lisp.net/project/slime/)/REPL-style workflow

Light Table is an IDE for Clojure, Python, and Javascript that is focused on providing many ways of evaluating and interacting with a running instance of your program, changing functions and variables to change the program's behavior while it is running. This is similar to how you code in Common Lisp when using SLIME, and Emacs Lisp in Emacs.

 * **Advantages**: As was mentioned by @CrossR, oni1 experimented with integrating Light Table-like code execution.  The Light Table model is advantageous in that it is tightly integrated into the editor and works both for notebook-style development and other, more traditional workflows (such as the development of backend servers and games). It could also allow for useful tools such as in-line variable watch. Possibly, the vim/neovim plugin [neoterm](https://github.com/kassio/neoterm) could be leveraged for this approach as well, to send code to running REPLs et cetera.

 * **Challenges**: As @bryphe mentioned, a generic model for in-line code execution is difficult to develop. Some REPL-based languages (lisps, Python, Haskell, R, Julia) are well suited, as evidenced by their prevalence in tools such as the mentioned Light Table, [Jupyter Notebook/Lab](https://jupyter.org/), RStudio, and the Emacs modes SLIME and [CIDER](https://github.com/clojure-emacs/cider). However, other languages, notably statically compiled ones with long compilation times such as C++ and Rust, are less suited, as evaluating a changed piece of code requires a recompile. Many such languages have [interactive code playgrounds](https://play.rust-lang.org/) which may be of use, though, and tree-sitter could possibly be leveraged to resolve variables outside the scope of the evaluated code block.

#### [Notebook](https://mybinder.org/v2/gh/ipython/ipython-in-depth/master?filepath=binder/Index.ipynb)/[Org-mode](https://orgmode.org/) workflow

The Notebook workflow differs from the REPL-based workflow in that the code is more like a linear document, to be read from start to finish. It has become ubiquitous among scientists (both data and otherwise), and Jupyter Notebooks support a whole bunch of languages. The user writes code and prose intertwined, and can evaluate blocks of code to provide inline results such as tables, images, and animations, or the more traditional output to stdout.

 * **Advantages**: Integrating with a Jupyter kernel could possibly be the easiest way to support in-line code execution, exploratory programming, and literate programming, where we could leverage the multitude of existing kernels to do most of the heavy lifting. Other, similar notebooks are [Observable](https://observablehq.com/) and [Zeppelin](https://zeppelin.apache.org/), though others will need to comment on how they differ from Jupyter as I have no experience with them.

 * **Disadvantages**: Notebooks are a different paradigm from more traditional programming approaches and don't integrate as easily into other types of workflows. They are also in binary format, and not as useful as pure text files when used with git, grep et cetera. They are also limited in their generalizability: A single notebook cannot contain multiple programming languages. Emacs' Org-mode[ solves both of these issues](https://orgmode.org/worg/org-contrib/babel/), but is much more niche and would require implementing a significant amount of the Emacs Org functionality implemented.

### TL:DR

It would be nice (dare I say a killer feature, at least compared to terminal vim) to have some sort of interactive workflow in onivim2. There are two approaches that I've seen in practice, the notebook approach and the REPL-like approach, each with advantages and disadvantages. If you know of other ways such features could be implemented, please add a comment!
